bigtable/google/cloud/bigtable/table.py

```diff
@@ -532,8 +544,48 @@ def _do_mutate_retryable_rows(self):
         return self.responses_statuses
 
 
+class _RetryableReadRows(object):
+    """A callable worker that can retry to read rows with transient errors.
+
+    This class is a callable that can retry reading rows that result in
+    transient errors.
+    """
+
+    def __init__(self, client, table_name, last_scanned_key,
+                 end_key, filter_, limit):
+        self.client = client
+        self.table_name = table_name
+        self.last_scanned_key = last_scanned_key
+        self.end_key = end_key
+        self.filter_ = filter_
+        self.limit = limit
+
+    def __call__(self):
+        from google.cloud.bigtable.row import _retry_commit_exception
+
+        read_rows = functools.partial(self._do_read_retryable_rows)
+        retry_ = Retry(
+            predicate=_retry_commit_exception,
+            deadline=30)
+        return retry_(read_rows)()
+
+    def _do_read_retryable_rows(self):
```

Comment:
Hrm, after reading this I'm not convinced this class needs to exist as all. Could it just be a function applied with functools.partial?

Reply:
This comment still applies.

Reply:
To remove the issues with throwing an exception inside a generator, which breaks the generator, retry was moved INSIDE of YieldRowsData.read_rows.  When the iterator on ReadRowsResponses throws an exception, the iterator is recreated using the value of last_scanned_row_key.  See the new commit for details. 

There could be one final refactoring, to use Retry for the following:
```
response = six.next(self._response_iterator)
```
To do so would introduce a method on YieldRowsData, which would have access to the state information needed to recreate the iterator (self.request_pb and self.client).  However, this should be run by @jonparrott first, since his preference is to create a simple functools.partial, as was done for MutateRow.  In our case, since we need to modify the request_pb with the last_scanned_row_key, this extra state information needs to be conveyed. The saved client is also needed to create the new iterator.  Please approve this design first.

Reply:
@jonparrott The last refactoring was successfully applied, so Retry is being employed to do the retries on the ReadRowsReponse iterator.  The solution is now fully encapsulated within YieldRowsData, which updates the last_scanned_row_key, which is needed for starting the retry read rows from the last successful scan point.  Ready for full review now.